src/ip_resolver.hpp

```diff
@@ -0,0 +1,77 @@
+/*
+    Copyright (c) 2007-2018 Contributors as noted in the AUTHORS file
+
+    This file is part of libzmq, the ZeroMQ core engine in C++.
+
+    libzmq is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    As a special exception, the Contributors give you permission to link
+    this library with independent modules to produce an executable,
+    regardless of the license terms of these independent modules, and to
+    copy and distribute the resulting executable under terms of your choice,
+    provided that you also meet, for each linked independent module, the
+    terms and conditions of the license of that module. An independent
+    module is a module which is not derived from or based on this library.
+    If you modify this library, you must extend this exception to your
+    version of the library.
+
+    libzmq is distributed in the hope that it will be useful, but WITHOUT
+    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+    FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+    License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef __ZMQ_IP_RESOLVER_HPP_INCLUDED__
+#define __ZMQ_IP_RESOLVER_HPP_INCLUDED__
+
+#if !defined ZMQ_HAVE_WINDOWS
+#include <sys/socket.h>
+#include <netinet/in.h>
+#endif
+
+namespace zmq
+{
+union ip_addr_t
+{
+    sockaddr generic;
+    sockaddr_in ipv4;
+    sockaddr_in6 ipv6;
+};
+
+class ip_resolver_t
+{
+  public:
+    ip_resolver_t ();
+
+    ip_resolver_t &bindable (bool bindable_);
```

Comment:
It would greatly reduce complexity if an ip_resolver_t were not reconfigurable. This is not necessary in the production code, and also not in the tests if the test cases are split up.

The class can be split up into a builder ip_resolver_builder_t (which contains the functions from lines 52-56, and an additional build function that creates an ip_resolver_t) and an ip_resolver_t (which contains the remaining functions, and gets a constructor that accepts all configuration parameters, and maybe an interface containing the functions that are platform-specific resp. intended to be mocked in the tests).

Reply:
Ok, I'm going to split the builder code and the tests.

For the test "overrides" you prefer adding an API to change the implementation instead of a linked overload? @bluca and you seem to have a slightly different approach here, I'm not sure what to do.

Reply:
Hm, from the comments here I cannot deduce what @bluca suggested, is it something you discussed on IRC?

I try to make my suggestion a bit more concrete, to ensure we talk about the same thing.
I imagine that ip_resolver_t gets a ctor such as
```
ip_resolver_t (..., i_dns_t);
```

with 
```
class i_dns_t
{
   virtual int getaddrinfo (const char *node_, const char *service_,
                     const struct addrinfo *hints_,
                     struct addrinfo **res_) = 0;
   virtual void freeaddrinfo (struct addrinfo *res_) = 0;

   virtual ~i_dns_t();
};
```
Maybe i_dns_t needs to contain other functions that need to be mock-injected by the tests. Maybe the signature of getaddrinfo can be abstracted somehow, such that it more closely fits the actual use in ip_resolver_t, and is somewhat more C++-style, to avoid the need for a separate freeaddrinfo function in the interface.

I can give it a closer look to make a more specific suggestion on this if you like.